kql/dsl.py

```diff
@@ -0,0 +1,116 @@
+                        if "should" not in dsl:
+                            dsl[child_type].extend(child_terms)
+                        else:
+                            dsl[boolean_type].append(boolean(should=child_terms))
+                    elif child_type == "must_not":
+                        dsl[child_type].extend(child_terms)
+                    elif child_type != "minimum_should_match":
+                        raise ValueError("Unknown term {}: {}".format(child_type, child_terms))
+            else:
+                dsl[boolean_type].append(child)
+
+    elif boolean_type == "should":
+        # can flatten `should` of `should`
+        for child in children:
+            if list(child) == ["bool"] and set(child["bool"]).issubset({"should", "minimum_should_match"}):
+                dsl["should"].extend(child["bool"]["should"])
+            else:
+                dsl[boolean_type].append(child)
+
+    elif boolean_type == "must_not" and len(children) == 1:
+        # must_not: [{bool: {must: x}}] -> {must_not: x}
+        child = children[0]
+        if list(child) == ["bool"] and list(child["bool"]) in (["filter", "must"]):
+            dsl = {"must_not": children[0]["must"]}
+        else:
+            dsl = {"must_not": children}
+
+    else:
+        dsl = dict(kwargs)
+
+    if "should" in dsl:
+        dsl.update(minimum_should_match=1)
+
+    dsl = {"bool": dict(dsl)}
+    return dsl
+
+
+class ToDsl(Walker):
+
+    def _walk_default(self, node, *args, **kwargs):
+        raise KqlCompileError("Unable to convert {}".format(node))
+
+    def _walk_exists(self, _):
+        return lambda field: {"exists": {"field": field}}
+
+    def _walk_wildcard(self, tree):
+        return lambda field: {"query_string": {"fields": [field], "query": tree.value}}
+
+    def _walk_value(self, tree):
+        return lambda field: {"match": {field: tree.value}}
+
+    def _walk_field(self, field):
+        return field.name
+
+    def _walk_field_range(self, tree):
+        operator_map = {"<": "lt", "<=": "lte", ">=": "gte", ">": "gt"}
+        field = self.walk(tree.field)
+        return {"range": {field: {operator_map[tree.operator]: tree.value.value}}}
+
+    def _walk_not_expr(self, tree):
+        return boolean(must_not=[self.walk(tree.expr)])
+
+    def _walk_and_expr(self, tree):
+        return boolean(filter=[self.walk(node) for node in tree.items])
+
+    def _walk_or_expr(self, tree):
+        return boolean(should=[self.walk(node) for node in tree.items])
+
+    def _walk_and_values(self, tree):
+        children = [self.walk(node) for node in tree.items]
+        return lambda field: boolean(filter=[child(field) for child in children])
+
+    def _walk_or_values(self, tree):
+        children = [self.walk(node) for node in tree.items]
+        return lambda field: boolean(should=[child(field) for child in children])
+
+    def _walk_not_value(self, tree):
+        child = self.walk(tree.expr)
```

Comment:
I looped through all of the rules to test with their query logic - the only errors were around `not`, since within scope of another field, can sometimes seem to not pass the `expr`

```
'NotValue' object has no attribute 'expr'
Connection to Internal Network via Telnet: event.category:network and event.type:(connection or start) and process.name:telnet and destination.ip:((10.0.0.0/8 or 172.16.0.0/12 or 192.168.0.0/16 or "FE80::/10") and not (127.0.0.0/8 or "::1/128"))

'NotValue' object has no attribute 'expr'
Elastic Endpoint: event.kind:alert and event.module:(endpoint and not endgame)
```